File: ios/Sources/RynBridgeAuthKakao/KakaoAuthProvider.swift
import Foundation
import RynBridge
import RynBridgeAuth

#if canImport(KakaoSDKUser)
import KakaoSDKCommon
import KakaoSDKAuth
import KakaoSDKUser

@available(iOS 17.0, *)
public final class KakaoAuthProvider: AuthProvider, @unchecked Sendable {
    private var currentToken: String?
    private var currentUser: AuthUser?

    public init() {
        self.currentToken = nil
        self.currentUser = nil
    }

    public func login(provider: String, scopes: [String]) async throws -> LoginResult {
        let token = try await loginWithKakao()
        currentToken = token.accessToken

        let user = try await fetchUser()
        currentUser = user

        let expiresAt = token.expiredAt.map { ISO8601DateFormatter().string(from: $0) }
            ?? ISO8601DateFormatter().string(from: Date().addingTimeInterval(3600))

        return LoginResult(
            token: token.accessToken,
            refreshToken: token.refreshToken,
            expiresAt: expiresAt,
            user: user
        )
    }

    private func loginWithKakao() async throws -> OAuthToken {
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.main.async {
                if UserApi.isKakaoTalkLoginAvailable() {
                    UserApi.shared.loginWithKakaoTalk { token, error in
                        if let error {
                            // Fallback to Kakao Account login
                            UserApi.shared.loginWithKakaoAccount { token, error in
                                if let error { continuation.resume(throwing: error) }
                                else if let token { continuation.resume(returning: token) }
                                else { continuation.resume(throwing: RynBridgeError(code: .unknown, message: "Kakao login failed")) }
                            }
                        } else if let token {
                            continuation.resume(returning: token)
                        } else {
                            continuation.resume(throwing: RynBridgeError(code: .unknown, message: "Kakao login failed"))
                        }
                    }
                } else {
                    UserApi.shared.loginWithKakaoAccount { token, error in
                        if let error { continuation.resume(throwing: error) }
                        else if let token { continuation.resume(returning: token) }
                        else { continuation.resume(throwing: RynBridgeError(code: .unknown, message: "Kakao login failed")) }
                    }
                }
            }
        }
    }

    private func fetchUser() async throws -> AuthUser? {
        return try await withCheckedThrowingContinuation { continuation in
            UserApi.shared.me { user, error in
                if let error {
                    continuation.resume(throwing: error)
                    return
                }
                guard let user else {
                    continuation.resume(returning: nil)
                    return
                }
                let authUser = AuthUser(
                    id: String(user.id ?? 0),
                    email: user.kakaoAccount?.email,
                    name: user.kakaoAccount?.profile?.nickname,
                    profileImage: user.kakaoAccount?.profile?.thumbnailImageUrl?.absoluteString
                )
                continuation.resume(returning: authUser)
            }
        }
    }

    public func logout() async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            UserApi.shared.logout { error in
                if let error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            }
        }
        currentToken = nil
        currentUser = nil
    }

    public func getToken() async throws -> TokenResult {
        return TokenResult(token: currentToken, expiresAt: nil)
    }

    public func refreshToken() async throws -> LoginResult {
        return try await login(provider: "kakao", scopes: [])
    }

    public func getUser() async throws -> AuthUser? {
        if let currentUser { return currentUser }
        return try await fetchUser()
    }
}
#endif

File: android/auth-kakao/src/main/kotlin/io/rynbridge/auth/kakao/KakaoAuthProvider.kt
package io.rynbridge.auth.kakao

import android.content.Context
import com.kakao.sdk.auth.model.OAuthToken
import com.kakao.sdk.common.model.ClientError
import com.kakao.sdk.common.model.ClientErrorCause
import com.kakao.sdk.user.UserApiClient
import io.rynbridge.auth.*
import kotlinx.coroutines.suspendCancellableCoroutine
import java.time.Instant
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Kakao Login Provider.
 *
 * Requires KakaoSDK initialization in Application.onCreate:
 *   KakaoSdk.init(this, "YOUR_NATIVE_APP_KEY")
 *
 * @param context Activity or Application context
 */
class KakaoAuthProvider(
    private val context: Context
) : AuthProvider {

    private var currentToken: String? = null
    private var currentUser: AuthUser? = null

    override suspend fun login(provider: String, scopes: List<String>): LoginResult {
        val token = loginWithKakao()
        currentToken = token.accessToken

        val user = fetchUser()
        currentUser = user

        return LoginResult(
            token = token.accessToken,
            refreshToken = token.refreshToken,
            expiresAt = token.accessTokenExpiresAt?.let { Instant.ofEpochMilli(it.time).toString() }
                ?: Instant.now().plusSeconds(3600).toString(),
            user = user
        )
    }

    private suspend fun loginWithKakao(): OAuthToken {
        return suspendCancellableCoroutine { cont ->
            val callback: (OAuthToken?, Throwable?) -> Unit = { token, error ->
                if (error != null) {
                    cont.resumeWithException(error)
                } else if (token != null) {
                    cont.resume(token)
                } else {
                    cont.resumeWithException(RuntimeException("Kakao login failed: no token"))
                }
            }

            if (UserApiClient.instance.isKakaoTalkLoginAvailable(context)) {
                UserApiClient.instance.loginWithKakaoTalk(context) { token, error ->
                    if (error != null) {
                        if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
                            cont.resumeWithException(error)
                            return@loginWithKakaoTalk
                        }
                        UserApiClient.instance.loginWithKakaoAccount(context, callback = callback)
                    } else {
                        callback(token, null)
                    }
                }
            } else {
                UserApiClient.instance.loginWithKakaoAccount(context, callback = callback)
            }
        }
    }

    private suspend fun fetchUser(): AuthUser? {
        return suspendCancellableCoroutine { cont ->
            UserApiClient.instance.me { user, error ->
                if (error != null) {
                    cont.resume(null)
                } else {
                    cont.resume(user?.let {
                        AuthUser(
                            id = it.id?.toString() ?: "",
                            email = it.kakaoAccount?.email,
                            name = it.kakaoAccount?.profile?.nickname,
                            profileImage = it.kakaoAccount?.profile?.thumbnailImageUrl
                        )
                    })
                }
            }
        }
    }

    override suspend fun logout() {
        suspendCancellableCoroutine { cont ->
            UserApiClient.instance.logout { error ->
                if (error != null) {
                    cont.resumeWithException(error)
                } else {
                    currentToken = null
                    currentUser = null
                    cont.resume(Unit)
                }
            }
        }
    }

    override suspend fun getToken(): TokenResult {
        return TokenResult(token = currentToken, expiresAt = null)
    }

    override suspend fun refreshToken(): LoginResult {
        return login("kakao", emptyList())
    }

    override suspend fun getUser(): AuthUser? {
        return currentUser ?: fetchUser()
    }
}
